File: workflow-engine/frontend/src/app/services/tramite.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { Tramite, PendingTask } from '../models/tramite.model';

@Injectable({
  providedIn: 'root'
})
export class TramiteService {

  private readonly API_URL = 'http://localhost:8080/api/tramites';

  constructor(private http: HttpClient) {}

  startTramite(policyId: string, userId: string, userName: string): Observable<Tramite> {
    return this.http.post<Tramite>(`${this.API_URL}/start`, { policyId, userId, userName });
  }

  completeTask(tramiteId: string, nodeId: string, userId: string, userName: string,
               formData: Record<string, any>, comment: string): Observable<Tramite> {
    return this.http.post<Tramite>(`${this.API_URL}/${tramiteId}/complete`, {
      nodeId, userId, userName, formData, comment
    });
  }

  getAll(): Observable<Tramite[]> {
    return this.http.get<Tramite[]>(this.API_URL);
  }

  getById(id: string): Observable<Tramite> {
    return this.http.get<Tramite>(`${this.API_URL}/${id}`);
  }

  getByStatus(status: string): Observable<Tramite[]> {
    return this.http.get<Tramite[]>(`${this.API_URL}/status/${status}`);
  }

  getByUser(userId: string): Observable<Tramite[]> {
    return this.http.get<Tramite[]>(`${this.API_URL}/user/${userId}`);
  }

  getPendingTasks(departmentId: string): Observable<PendingTask[]> {
    return this.http.get<PendingTask[]>(`${this.API_URL}/tasks/department/${departmentId}`);
  }

  cancel(id: string): Observable<Tramite> {
    return this.http.patch<Tramite>(`${this.API_URL}/${id}/cancel`, {});
  }
}

File: workflow-engine/mobile/lib/services/tramite_service.dart
import 'dart:convert';
import '../models/tramite.dart';
import 'api_client.dart';

class TramiteService {
  static Future<List<Tramite>> getAll() async {
    final response = await ApiClient.get('/tramites');
    if (response.statusCode == 200) {
      final List<dynamic> data = jsonDecode(response.body);
      return data.map((t) => Tramite.fromJson(t)).toList();
    }
    throw Exception('Error al obtener trámites');
  }

  static Future<Tramite> getById(String id) async {
    final response = await ApiClient.get('/tramites/$id');
    if (response.statusCode == 200) {
      return Tramite.fromJson(jsonDecode(response.body));
    }
    throw Exception('Trámite no encontrado');
  }

  static Future<List<Tramite>> getByUser(String userId) async {
    final response = await ApiClient.get('/tramites/user/$userId');
    if (response.statusCode == 200) {
      final List<dynamic> data = jsonDecode(response.body);
      return data.map((t) => Tramite.fromJson(t)).toList();
    }
    throw Exception('Error al obtener trámites del usuario');
  }

  static Future<List<Tramite>> getByStatus(String status) async {
    final response = await ApiClient.get('/tramites/status/$status');
    if (response.statusCode == 200) {
      final List<dynamic> data = jsonDecode(response.body);
      return data.map((t) => Tramite.fromJson(t)).toList();
    }
    throw Exception('Error al obtener trámites por estado');
  }

  static Future<List<PendingTask>> getPendingTasks(String departmentId) async {
    final response = await ApiClient.get('/tramites/tasks/department/$departmentId');
    if (response.statusCode == 200) {
      final List<dynamic> data = jsonDecode(response.body);
      return data.map((t) => PendingTask.fromJson(t)).toList();
    }
    throw Exception('Error al obtener tareas pendientes');
  }

  /// Obtiene MIS tareas (pendientes + completadas) filtradas por mi departamento.
  /// Usa el endpoint del backend que ya hace el filtro correcto.
  static Future<Map<String, dynamic>> getMyTasks(String departmentId) async {
    final response = await ApiClient.get('/tramites/tasks/user/$departmentId');
    if (response.statusCode == 200) {
      return Map<String, dynamic>.from(jsonDecode(response.body));
    }
    throw Exception('Error al obtener mis tareas');
  }

  static Future<Tramite> completeTask({
    required String tramiteId,
    required String nodeId,
    required String userId,
    required String userName,
    required Map<String, dynamic> formData,
    required String comment,
  }) async {
    final response = await ApiClient.post('/tramites/$tramiteId/complete', {
      'nodeId': nodeId,
      'userId': userId,
      'userName': userName,
      'formData': formData,
      'comment': comment,
    });

    if (response.statusCode == 200) {
      return Tramite.fromJson(jsonDecode(response.body));
    }
    final error = jsonDecode(response.body);
    throw Exception(error['error'] ?? 'Error al completar tarea');
  }

  static Future<Tramite> startTramite({
    required String policyId,
    required String userId,
    required String userName,
  }) async {
    final response = await ApiClient.post('/tramites/start', {
      'policyId': policyId,
      'userId': userId,
      'userName': userName,
    });

    if (response.statusCode == 201) {
      return Tramite.fromJson(jsonDecode(response.body));
    }
    final error = jsonDecode(response.body);
    throw Exception(error['error'] ?? 'Error al iniciar trámite');
  }
}
